libraries-3/src/main/java/com/baeldung/jnats/NatsClient.java

```diff
@@ -1,123 +1,144 @@
+    static class CustomErrorListener implements ErrorListener {
+        @Override
+        public void errorOccurred(Connection conn, String error) {
+            log.error("Error Occurred: {}", error);
+        }
+
+        @Override
+        public void exceptionOccurred(Connection conn, Exception exp) {
+            log.error("Exception Occurred: {}", exp.toString());
+        }
+    }
+    private void initConnection() {
+        try {
+            Options options = new Options.Builder()
+                .server(serverURI)
+                .connectionListener((connection, event) -> log.info("Connection Event: {}", event.toString()))
+                .errorListener(new CustomErrorListener())
+                .build();
+            natsConnection = Nats.connect(options);
+        } catch (IOException | InterruptedException e) {
+            log.error("Error connecting to NATs! {}", e.toString());
+        }
     }
 
     public void closeConnection() {
         // Close connection
-        natsConnection.close();
+        try {
+            natsConnection.close();
+        }
+        catch (InterruptedException e) {
+            log.warn("Warning:", e);
+        }
     }
 
-    private Connection initConnection(String uri) {
+    void publishMessage(String subject, String message) {
         try {
-            Options options = new Options.Builder()
-              .errorCb(ex -> log.error("Connection Exception: ", ex))
-              .disconnectedCb(event -> log.error("Channel disconnected: {}", event.getConnection()))
-              .reconnectedCb(event -> log.error("Reconnected to server: {}", event.getConnection()))
-              .build();
-
-            return Nats.connect(uri, options);
-        } catch (IOException ioe) {
-            log.error("Error connecting to NATs! ", ioe);
-            return null;
+            natsConnection.publish(subject, message.getBytes());
+        } catch (IllegalStateException ise) {
+            log.error("Error publishing message: {} to {} ", message, subject, ise);
         }
     }
 
-    void publishMessage(String topic, String replyTo, String message) {
+    void publishMessageWithReply(String subject, String replyTo, String message) {
         try {
-            natsConnection.publish(topic, replyTo, message.getBytes());
-        } catch (IOException ioe) {
-            log.error("Error publishing message: {} to {} ", message, topic, ioe);
+            natsConnection.publish(subject, replyTo, message.getBytes());
+        } catch (IllegalStateException ise) {
+            log.error("Error publishing message: {} to {} ", message, subject, ise);
         }
     }
 
-    public void subscribeAsync(String topic) {
-
-        AsyncSubscription subscription = natsConnection.subscribe(
-          topic, msg -> log.info("Received message on {}", msg.getSubject()));
-
+    public Subscription subscribeAsync(String subject) {
+        Dispatcher dispatcher = natsConnection.createDispatcher();
```

Comment:
What is the lifecycle of a Disptacher? When should it be closed?

Reply:
It will be close when the connection is closed. Dispatcher is always created and maintained from the connection and the connection is AutoCloseable.

Reply:
I made the client AutoCloseable